Mathlib/Lean/Meta/KAbstractPositions.lean

```diff
@@ -0,0 +1,88 @@
+/-
+Copyright (c) 2023 Jovan Gerbscheid. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jovan Gerbscheid
+-/
+import Lean.HeadIndex
+import Lean.Meta.ExprLens
+import Lean.Meta.Check
+
+/-!
+
+# Find the positions of a pattern in an expression
+
+This file defines some tools for dealing with sub expressions and occurrence numbers.
+This is used for creating a `rw` tactic call that rewrites a selected expression.
+
+`viewKAbstractSubExpr` takes an expression and a position in the expression, and returns
+the sub-expression together with an optional occurrence number that would be required to find
+the sub-expression using `kabstract` (which is what `rw` uses to find the position of the rewrite)
+
+`rw` can fail if the motive is not type correct. `kabstractIsTypeCorrect` checks
+whether this is the case.
+
+-/
+
+namespace Lean.Meta
+
+/-- Return the positions that `kabstract` would abstract for pattern `p` in expression `e`.
+i.e. the positions that unify with `p`. -/
+def kabstractPositions (p e : Expr) : MetaM (Array SubExpr.Pos) := do
+  let e ← instantiateMVars e
+  let mctx ← getMCtx
+  let pHeadIdx := p.toHeadIndex
+  let pNumArgs := p.headNumArgs
+  let rec
+  /-- The main loop that loops though all subexpressions -/
+  visit (e : Expr) (pos : SubExpr.Pos) (positions : Array SubExpr.Pos) :
+      MetaM (Array SubExpr.Pos) := do
+    let visitChildren : Array SubExpr.Pos → MetaM (Array SubExpr.Pos) :=
+      match e with
+      | .app f a         => visit f pos.pushAppFn
+                        >=> visit a pos.pushAppArg
+      | .mdata _ b       => visit b pos
+      | .proj _ _ b      => visit b pos.pushProj
+      | .letE _ t v b _  => visit t pos.pushLetVarType
+                        >=> visit v pos.pushLetValue
+                        >=> visit b pos.pushLetBody
+      | .lam _ d b _     => visit d pos.pushBindingDomain
+                        >=> visit b pos.pushBindingBody
+      | .forallE _ d b _ => visit d pos.pushBindingDomain
+                        >=> visit b pos.pushBindingBody
+      | _                => pure
+    if e.hasLooseBVars then
+      visitChildren positions
+    else if e.toHeadIndex != pHeadIdx || e.headNumArgs != pNumArgs then
+      visitChildren positions
+    else
+      if (← isDefEq e p) then
+        setMCtx mctx
```

Comment:
I don't understand what `setMCtx` is for since it looks to me that `mctx` is defined only once outside the loop and never modified. Can you quickly explain?

Reply:
Yes, `isDefEq` can change the `mctx` if it succeeds. In particular if you run `isDefEq` on `?m =?= 5`, then after it returns `true`, the assignment `?m := 5` is added to the `MetavarContext`.

The (new) behaviour of `kabstract` is that it only keeps the `MetavarContext` for expressions that are abstracted, and not the ones that are skipped. This means that it is possible to rewrite in either of the two positions of `1+2+3` using `add_comm`, by setting the position to be either 1 or 2. Resetting the `MetavarContext` is important, because if we choose position 2, we want to forget the assignments of `?a` and `?b` after unifying `?a + ?b` with `1 + 2 + 3`, so that later we can unify it with `1 + 2`.

Reply:
I see, thank you! Didn't know how `isDefEq` works.

Maybe a comment like ``-- reset `mctx` as `isDefEq` might change the context`` would help somebody like me?